src/neo/Ledger/Blockchain.cs

```diff
@@ -237,9 +273,34 @@ private VerifyResult OnNewBlock(Block block)
                 // We can store the new block in block_cache and tell the new height to other nodes after Persist().
                 system.LocalNode.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(Singleton.Height)));
             }
+            else
+            {
+                if (block.Index + 99 >= HeaderHeight)
+                    system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = block });
+                if (block.Index == HeaderHeight + 1)
+                {
+                    Header header = block.Header;
+                    recentHeaders.Add(header);
+                }
+            }
             return VerifyResult.Succeed;
         }
 
+        private void OnNewHeaders(Header[] headers)
+        {
+            using (SnapshotCache snapshot = GetSnapshot())
+            {
+                foreach (Header header in headers)
+                {
+                    if (header.Index > HeaderHeight + 1) break;
+                    if (header.Index < HeaderHeight + 1) continue;
+                    if (!header.Verify(snapshot)) break;
+                    recentHeaders.Add(header);
```

Comment:
If the cache is full, we should throw away the extra headers. Otherwise, it will overwrite the headers that has not yet been written to storage.

Reply:
fixed